### PR TITLE
UITableView + UIRefreshControl

### PR DESCRIPTION
**Pull request by paulofierro**
Added demo code that shows:

- how to set up and customize a `UIRefreshControl` with a table view
- render custom header views for a table view
- render custom cells for a table view which auto-size depending on the contents
- handle tapping on cells
- handle deleting cells in 3 different ways, depending on iOS version
- basic drag support (internal and external)
- basic drop support (internal and external)

### CODE REVIEW

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Paulo/Accessory Views/Cell/PauloTableViewCell.swift [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+            return
+        }
+
+        activityIndicator.startAnimating()
+
+        // ⚠️ THIS IS AWFUL!!!
+        // This should never be shipped in a real app. Use
+        // a third-party library like Kingfisher, SDWebImage
+        // or similar, or roll your own code that downloads the
+        // image properly, in an asynchronous manner and caches it.
```
Comment (ivancr) [inline BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Paulo/Accessory Views/Cell/PauloTableViewCell.swift:72]: Thanks for this comment 👍🏼

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Paulo/Accessory Views/Header/PauloHeaderView.swift [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+            else { return nil}
+
+        let gradientLayer = CAGradientLayer()
+        gradientLayer.colors = [topColor.cgColor, bottomColor.cgColor]
+        gradientLayer.bounds = bounds
+        UIGraphicsBeginImageContextWithOptions(gradientLayer.bounds.size, false, 0.0)
+
+        defer {
+            UIGraphicsEndImageContext()
+        }
```
Comment (ivancr) [inline BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Paulo/Accessory Views/Header/PauloHeaderView.swift:32]: nice

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Paulo/PauloViewController.swift [hunk 3]
```diff
@@ -6,13 +6,340 @@
...
+        let cellNib = UINib(nibName: String(describing: PauloTableViewCell.self), bundle: nil)
+        tableView.register(cellNib, forCellReuseIdentifier: PauloTableViewCell.identifier)
+
+        // Setting this gives us a slight performance boost, from the docs:
+        //   "using estimation allows you to defer some of the cost of geometry
+        //    calculation from load time to scrolling time."
+        tableView.estimatedRowHeight = PauloTableViewCell.estimatedRowHeight
+
+        // Center cells on iPad so its a bit more usable
+        tableView.cellLayoutMarginsFollowReadableWidth = true
```
Comment (ivancr) [inline BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Paulo/PauloViewController.swift:51]: This looks familiar :)

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Paulo/PauloViewController.swift [hunk 4]
```diff
@@ -6,13 +6,340 @@
...
+        let headerNib = UINib(nibName: String(describing: PauloHeaderView.self), bundle: nil)
+        tableView.register(headerNib, forHeaderFooterViewReuseIdentifier: PauloHeaderView.identifier)
+
+        let cellNib = UINib(nibName: String(describing: PauloTableViewCell.self), bundle: nil)
+        tableView.register(cellNib, forCellReuseIdentifier: PauloTableViewCell.identifier)
+
+        // Setting this gives us a slight performance boost, from the docs:
+        //   "using estimation allows you to defer some of the cost of geometry
+        //    calculation from load time to scrolling time."
+        tableView.estimatedRowHeight = PauloTableViewCell.estimatedRowHeight
```
Comment (ivancr) [inline BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Paulo/PauloViewController.swift:48]: I recently posted on Slack an excellent talk about self sizing cells, did you watch it?

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Paulo/PauloViewController.swift [hunk 5]
```diff
@@ -6,13 +6,340 @@
...
+
+        // Center cells on iPad so its a bit more usable
+        tableView.cellLayoutMarginsFollowReadableWidth = true
+
+        // Set up delegates (the first two can be done in IB)
+        tableView.dataSource = self
+        tableView.delegate = self
+        tableView.dragDelegate = self
+        tableView.dropDelegate = self
```
Comment (ivancr) [inline BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Paulo/PauloViewController.swift:57]: fancyyyyyyy!

File: BackToBasicsUIKit/BackToBasicsUIKit/MainTableViewController.swift [hunk 6]
```diff
@@ -51,10 +51,10 @@ class MainTableViewController: UITableViewController {
 
         return cell
     }
-    
+
     override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
         var demoVC = UIViewController()
-        
+
```
Comment (ivancr) [inline BackToBasicsUIKit/BackToBasicsUIKit/MainTableViewController.swift:57]: 😀

**Review by ivancr — APPROVED** [review]
God damn! What a fantastic PR. You're going to need some time to go through all of this code. There are so many cool things here to discuss.
Amazing job man!

File: BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Paulo/PauloViewController.swift (hunk 4 above)
Comment (paulofierro) [reply BackToBasicsUIKit/BackToBasicsUIKit/Demo ViewControllers/Paulo/PauloViewController.swift:48]: I must have missed that!

**Pull request merged**